compendium/DeclarativeServices/test/TestServiceDependencyErrorLogging.cpp

```diff
@@ -0,0 +1,155 @@
+/*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  =============================================================================*/
+
+#include <gtest/gtest.h>
+
+#include "../src/manager/ComponentConfigurationImpl.hpp"
+#include "../src/manager/ReferenceManager.hpp"
+#include "../src/manager/SingletonComponentConfiguration.hpp"
+#include "../src/metadata/ComponentMetadata.hpp"
+#include "Mocks.hpp"
+#include "cppmicroservices/Framework.h"
+#include "cppmicroservices/FrameworkEvent.h"
+#include "cppmicroservices/FrameworkFactory.h"
+#include "cppmicroservices/ServiceInterface.h"
+
+#include <string>
+
+namespace cppmicroservices {
+namespace scrimpl {
+
+class CustomLogger : public FakeLogger
+{
+  std::string msg;
+
+public:
+  using FakeLogger::Log;
+  void Log(cppmicroservices::logservice::SeverityLevel,
+           const std::string& message) override
+  {
+    msg = message;
+  }
+
+  std::string GetLoggerMessage() { return msg; }
+};
+
+class ServiceDependencyErrorLoggingTest : public ::testing::Test
+{
+
+public:
+  std::shared_ptr<CustomLogger> GetLogger() { return logger_; }
+
+protected:
+  ServiceDependencyErrorLoggingTest()
+      : framework(cppmicroservices::FrameworkFactory().NewFramework())
+  {
+    logger_ = std::make_shared<CustomLogger>(); 
+  }
+
+  virtual ~ServiceDependencyErrorLoggingTest() = default;
+
+  virtual void SetUp()
+  {
+    framework.Start();
+  }
+
+  virtual void TearDown()
+  {
+    framework.Stop();
+    framework.WaitForStop(std::chrono::milliseconds::zero());
+  }
+
+ cppmicroservices::Framework& GetFramework() { return framework; }
+
+private:
+  std::shared_ptr<CustomLogger> logger_;
+
+
+public:
+  cppmicroservices::Framework framework;
+};
+
+
+TEST_F(ServiceDependencyErrorLoggingTest, TestServiceDependencyLDAPFilter)
```

Comment:
This seems like a complicated test for this new code. It creates several mock objects including a SingletonComponentConfigurationImpl object. Just wondering if a simpler test would have been sufficient. GetMatchingServiceListeners is called from ServiceRegistrationBase::SetProperties, ServiceRegistrationBase::Unregister and ServiceRegistrationBase::CreateServiceProperties. Just wondering if a test using a ServiceRegistrationBase object would have been sufficient. It would be simpler.

Reply:
`mockMetadata` and `refmockMetadata` creates dummy metadata values needed for the use case. i.e creating non-matching LDAP filter value to enable logging. `SingletonComponentConfigurationImpl` creates registration & reference manager that automatically handle the service registration in case of dependencies by triggering the call to Register at the appropriate time. 
I looked into `ServiceRegistrationBase` and it looks like it doesn't handle dependencies tracking and doesn't have a logger to verify what information is logged. We may have to handle all those cases manually.

Reply:
Jeff sent me a link to the Solution Review document for this feature [Solution Review : Improve service dependency error logging](https://confluence.mathworks.com/display/I18N/Solution+Review+%3A+Improve+service+dependency+error+logging)
That helped me understand what problem you are trying to solve and why the complexity is necessary. Looks good.